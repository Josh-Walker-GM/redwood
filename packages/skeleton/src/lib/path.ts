import path from 'path'

import { getBaseDir, getConfigPath } from '@redwoodjs/internal/dist/index'

/**
 * Determines the root path of a redwood project which contains the provided path
 * @param pathWithinProject A path within a redwood project
 * @returns The root path of the associated project
 */
export function getRootPath(pathWithinProject: string) {
  return pathWithinProject
    ? path.dirname(getConfigPath(pathWithinProject))
    : getBaseDir()
}
