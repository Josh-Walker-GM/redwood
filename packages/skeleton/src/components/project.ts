import fs from 'fs'
import path from 'path'

import chalk from 'chalk'

import { getPaths } from '@redwoodjs/internal/dist/index'

import { getRootPath } from '../lib/path'

import { RedwoodSkeleton } from './base'
import { extractCells, RedwoodCell } from './cell'
import { extractFunctions, RedwoodFunction } from './function'
import { extractLayouts, RedwoodLayout } from './layout'
import { extractPages, RedwoodPage } from './page'
import { extractRouters, RedwoodRouter } from './router'
import { extractSides, RedwoodSide } from './side'

/**
 * Used to enumerate either JS or TS project types
 */
export enum RedwoodProjectType {
  TYPESCRIPT = 'typescript',
  JAVASCRIPT = 'javascript',
}

export class RedwoodProject extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly type: RedwoodProjectType

  private sides?: RedwoodSide[] | undefined
  private cells?: RedwoodCell[] | undefined
  private routers?: RedwoodRouter[] | undefined
  private layouts?: RedwoodLayout[] | undefined
  private pages?: RedwoodPage[] | undefined
  private functions?: RedwoodFunction[] | undefined

  public static getProject({
    pathWithinProject = '',
    full = false,
    readFromCache = true,
    insertIntoCache = true,
  }: {
    pathWithinProject?: string
    full?: boolean
    readFromCache?: boolean
    insertIntoCache?: boolean
  } = {}) {
    const projectCache = RedwoodProjectsCache.getInstance().projects
    const rootPath = getRootPath(pathWithinProject)

    if (readFromCache) {
      const cachedProject = projectCache.get(rootPath)
      if (cachedProject) {
        return cachedProject
      }
    }

    const project = new RedwoodProject(rootPath, full)
    if (insertIntoCache) {
      projectCache.set(project.filepath, project)
    }
    return project
  }

  private constructor(rootPath: string, full = false) {
    super(rootPath)

    // A project is typescript if we detect a tsconfig.json
    this.type =
      fs.existsSync(
        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
      ) ||
      fs.existsSync(
        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
      )
        ? RedwoodProjectType.TYPESCRIPT
        : RedwoodProjectType.JAVASCRIPT

    if (full) {
      this.sides = extractSides(this)
      this.cells = extractCells(this)
      this.layouts = extractLayouts(this)
      this.pages = extractPages(this)
      this.functions = extractFunctions(this)
      this.routers = extractRouters(this)
    }
  }

  getSides(forceExtract = false): RedwoodSide[] {
    if (forceExtract) {
      return extractSides(this)
    }
    return this.sides === undefined ? extractSides(this) : this.sides
  }

  getCells(forceExtract = false): RedwoodCell[] {
    if (forceExtract) {
      return extractCells(this)
    }
    return this.cells === undefined ? extractCells(this) : this.cells
  }

  getRouters(forceExtract = false): RedwoodRouter[] {
    if (forceExtract) {
      return extractRouters(this)
    }
    return this.routers === undefined ? extractRouters(this) : this.routers
  }

  getLayouts(forceExtract = false): RedwoodLayout[] {
    if (forceExtract) {
      return extractLayouts(this)
    }
    return this.layouts === undefined ? extractLayouts(this) : this.layouts
  }

  getPages(forceExtract = false): RedwoodPage[] {
    if (forceExtract) {
      return extractPages(this)
    }
    return this.pages === undefined ? extractPages(this) : this.pages
  }

  getFunctions(forceExtract = false): RedwoodFunction[] {
    if (forceExtract) {
      return extractFunctions(this)
    }
    return this.functions === undefined
      ? extractFunctions(this)
      : this.functions
  }

  // Diagnostics

  getInformation(): string {
    return '' // TODO: Implement, gets the info of all other components too
  }

  printInformation(): void {
    console.log(
      `${chalk.bgCyan('[Info]')}\t${this.name}\t${chalk.dim(this.filepath)}`
    )
  }

  printWarnings(cascade = false): void {
    if (this.warnings.length > 0) {
      const titleLine = `${chalk.bgYellow('[Warn]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const warningLines = this.warnings.map((warning, index) => {
        return ` (${index + 1}) ${warning}\n`
      })
      console.log(titleLine.concat('\n', ...warningLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printWarnings()
      })
      this.functions?.forEach((func) => {
        func.printWarnings()
      })
      this.layouts?.forEach((layout) => {
        layout.printWarnings()
      })
      this.pages?.forEach((page) => {
        page.printWarnings()
      })
      this.routers?.forEach((router) => {
        router.printWarnings()
        router.routes.forEach((route) => {
          route.printWarnings()
        })
      })
      this.sides?.forEach((side) => {
        side.printWarnings()
      })
    }
  }

  printErrors(cascade = false): void {
    if (this.errors.length > 0) {
      const titleLine = `${chalk.bgRed('[Error]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const errorLines = this.errors.map((error, index) => {
        return ` (${index + 1}) ${error}\n`
      })
      console.log(titleLine.concat('\n', ...errorLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printErrors()
      })
      this.functions?.forEach((func) => {
        func.printErrors()
      })
      this.layouts?.forEach((layout) => {
        layout.printErrors()
      })
      this.pages?.forEach((page) => {
        page.printErrors()
      })
      this.routers?.forEach((router) => {
        router.printErrors()
        router.routes.forEach((route) => {
          route.printErrors()
        })
      })
      this.sides?.forEach((side) => {
        side.printErrors()
      })
    }
  }
}

/**
 * Singelton class which contains a cache of {@link RedwoodProject} instances.
 *
 * This class is not exported and should only be available from static methods within {@link RedwoodProject}, in order to keep the surface of the skeleton API cleaner and leaner.
 */
class RedwoodProjectsCache {
  private static instance: RedwoodProjectsCache

  public projects: Map<string, RedwoodProject>

  private constructor() {
    this.projects = new Map()
  }

  public static getInstance(): RedwoodProjectsCache {
    if (!RedwoodProjectsCache.instance) {
      RedwoodProjectsCache.instance = new RedwoodProjectsCache()
    }
    return RedwoodProjectsCache.instance
  }
}
