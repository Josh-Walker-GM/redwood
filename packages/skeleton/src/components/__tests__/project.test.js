import path from 'path'

import { RedwoodProject } from '../project'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)
describe.each([
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
])('From within the %s fixture', (PROJECT_NAME) => {
  const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = PROJECT_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  test('Minimal project matches expectation', () => {
    const project = RedwoodProject.getProject({
      readFromCache: false,
      insertIntoCache: false,
    })
    project.filepath = project.filepath.substring(PROJECT_PATH.length)
    expect(project).toMatchSnapshot()
  })

  test.todo('Full project matches expectation')
})
