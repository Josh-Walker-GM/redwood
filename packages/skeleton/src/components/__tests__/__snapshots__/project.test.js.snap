// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture Minimal project matches expectation 1`] = `
RedwoodProject {
  "cells": undefined,
  "errors": [],
  "filepath": "",
  "layouts": undefined,
  "name": "empty-project",
  "pages": undefined,
  "routers": undefined,
  "sides": undefined,
  "type": "typescript",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture Minimal project matches expectation 1`] = `
RedwoodProject {
  "cells": undefined,
  "errors": [],
  "filepath": "",
  "layouts": undefined,
  "name": "example-todo-main",
  "pages": undefined,
  "routers": undefined,
  "sides": undefined,
  "type": "typescript",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture Minimal project matches expectation 1`] = `
RedwoodProject {
  "cells": undefined,
  "errors": [],
  "filepath": "",
  "layouts": undefined,
  "name": "example-todo-main-with-errors",
  "pages": undefined,
  "routers": undefined,
  "sides": undefined,
  "type": "javascript",
  "warnings": [],
}
`;

exports[`From within the test-project fixture Minimal project matches expectation 1`] = `
RedwoodProject {
  "cells": undefined,
  "errors": [],
  "filepath": "",
  "layouts": undefined,
  "name": "test-project",
  "pages": undefined,
  "routers": undefined,
  "sides": undefined,
  "type": "typescript",
  "warnings": [],
}
`;
